askama/src/lib.rs

```diff
@@ -510,6 +510,21 @@ pub trait Template {
     }
     /// Renders the template to the given `writer` buffer
     fn render_into(&self, writer: &mut dyn std::fmt::Write) -> Result<()>;
+    /// Helper method which allocates a new `Vec<u8>` and renders into it
+    fn render_bytes(&self) -> Result<Vec<u8>> {
+        let mut buf = Vec::with_capacity(self.size_hint());
+        self.render_into_bytes(&mut buf)?;
+        Ok(buf)
+    }
+    /// Renders the template to the given `writer`. (It is recommended to use a buffered implementation.)
+    fn render_into_bytes(&self, writer: &mut dyn std::io::Write) -> Result<()> {
+        let mut adapter = shared::io::WriteIoToFmt::new(writer);
+        let outcome = self.render_into(&mut adapter);
+        if let Some(error) = adapter.done() {
+            return Err(Error::Io(error));
+        }
+        return outcome;
```

Comment:
How did you decide that the `io::Error` should take precedence over the `render_into()` error?

```suggestion
        match (adapter.done(), outcome) {
            (Some(err), _) => Err(err),
            (_, result) => result,
        }
```

Reply:
My reasoning is that an I/O error is far more important than a formatting error.  (For example if I have an application that writes a report on disk, then in case I have both a formatting error and an I/O error, the formatting error is most likely temporary and fixable, meanwhile an I/O error is something more serious and permanent, and an administrator should take a look.)

Moreover, having the two errors at the same time should be practically impossible (given the sequential nature of the code):
* if the formatting error happens first, the template code should exit without trying to write anything no the writer;
* if the writing error happens, then the template code should exit without trying to format anything further;

I've applied the suggested change.